kong/plugins/aws-lambda/handler.lua

```diff
@@ -25,32 +24,77 @@ function AWSLambdaHandler:new()
   AWSLambdaHandler.super.new(self, "aws-lambda")
 end
 
-local function retrieve_parameters()
-  ngx_req_read_body()
+local function retrieve_parameters(content_type, conf)
   local body_parameters, err
-  local content_type = ngx_req_get_headers()[CONTENT_TYPE]
-  if content_type and string_find(content_type:lower(), "multipart/form-data", nil, true) then
+  if content_type and string_find(content_type, "multipart/form-data", nil, true) then
     body_parameters = Multipart(ngx_req_get_body_data(), content_type):get_all()
-  elseif content_type and string_find(content_type:lower(), "application/json", nil, true) then
+  elseif content_type and string_find(content_type, "application/json", nil, true)
+         and not conf.request_body then
     body_parameters, err = cjson.decode(ngx_req_get_body_data())
     if err then
       body_parameters = {}
     end
-  else
+  elseif content_type and string_find(content_type, "application/x-www-form-urlencoded", nil, true) then
     body_parameters = public_utils.get_post_args()
+  else
+    body_parameters = {}
   end
 
   return utils.table_merge(ngx_req_get_uri_args(), body_parameters)
 end
 
+local function retrieve_body(content_type)
+  local err, body
+  local raw_body = ngx_req_get_body_data()
+  if raw_body == nil then return nil end
+
+  if content_type and string_find(content_type, "application/json", nil, true) then
+    body, err = cjson.decode(raw_body)
+    if err then
+      return nil
+    end
+  elseif content_type and ( string_find(content_type, "application/xml", nil, true)
+    or  string_find(content_type, "application/soap+xml", nil, true)
+    or  string_find(content_type, "text/", nil, true) )
+  then
+    body = raw_body
+  else
+    body = ngx_encode_base64(raw_body)
+  end
+  return body
+end
```

Comment:
The fallback is because most of the other content types are binary.
We didn't decode body when JSON in retrieve parameters in case when request_body is set in order to avoid decoding it twice. I'd say that in case of JSON it makes sense to forward JSON request body as part of JSON wrapper. Though it does not matter for me. Please look at the refactored PR.